### ISSUE TITLE
The URL of one blob snapshot shows incorrectly

### ISSUE THREAD

**Issue by v-limu** [microsoft/AzureStorageExplorer#2788]
**Storage Explorer Version:** 1.13.0
**Build**: [20200403.15](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=3608332)
**Branch**: rel/1.13.0
**Platform/OS:** Windows 10/ Linux Ubuntu 16.04/ macOS High Sierra
**Architecture**: ia32/x64
**Regression From:** Not a regression

**Steps to reproduce:**
1. Expand one Non- ADLS Gen2 storage account -> Blob Containers.
2. Create a new blob container -> Upload one blob to it -> Create snapshot for this blob -> Manage Snapshots.
3. Right click the blob snapshot -> Copy URL -> Paste the URL.
4. Check the pasted URL.

**Expect Experience:**
The URL shows correctly.
![image](https://user-images.githubusercontent.com/54055206/78342958-b34a7180-75cc-11ea-8242-9a98412e6d3c.png)

**Actual Experience:**
The URL shows incorrectly. 
![image](https://user-images.githubusercontent.com/54055206/78341376-2dc5c200-75ca-11ea-9188-d7e1561d35fc.png)

**More Info:**
1. This issue also reproduces for **one folder under one file share snapshot** in Properties dialog.
![image](https://user-images.githubusercontent.com/54055206/78343094-eab91e00-75cc-11ea-8bcd-126dd3f6147b.png)

**Comment by JasonYeMSFT** [comment]
The url is the url returned by the api. We won't fix it on our side.

**Comment by egdedudette** [comment]
I confirm that this is true. Not only the copy to URL button is gone but you can find it by right click on the file but it does give a local URL instead of the usable URL.
FIX IT ASAP PLEASE!

**Comment by JasonYeMSFT** [comment]
@egdedudette Thanks for raising this up. We will re-triage this issue.

**Comment by egdedudette** [comment]
Why was this closed?

**Comment by JasonYeMSFT** [comment]
Because the last time we triage this issue, we found that "the url is the url returned by the api" and this issue is raised by our tester. We usually don't process data directly from the server unless there is an explicit need.

**Comment by egdedudette** [comment]
There is a need because We use it to test files and now we can't. Big problem for us.

**Comment by egdedudette** [comment]
This was created by the new update.

**Comment by craxal** [comment]
@egdedudette 
> We use it to test files and now we can't.

Are you saying that this was working in a previous version? If so, can you share the version number that this was working in? Also, can you share examples of the URLs you are receiving from the previous version and the new version? We want to make sure we're understanding your problem so we know what exactly needs to be fixed.

**Comment by egdedudette** [comment]
OK. I am able to copy the URL on my office compute with CRD. It works there.  version 1.12.0  Azure 10.3.3
It gives a URL like the previous poster shows with your local file path instead of the coded URL.
I'd rather not post the path but it does add these 3 characters: 
%2F  
after the user name.

**Comment by craxal** [comment]
@edgedudette The %2F characters are the URL encoding for "/". CRD? User name? You mean something like https://user@host.com?

You don't have to share the exact URLs. You can edit the endpoints and filenames to something more generic. Just enough that we can get a sense of the URL form you're getting and the form you're expecting.

**Comment by egdedudette** [comment]
https://curriculumtesting.blob.core.windows.net/files/FolderName%2Fnameoffnextfolderr%2Findex.html

Actually I just noticed it is in 2 places. After the folder name and between the file name.

**Comment by craxal** [comment]
@egdedudette I think that's by design. The encoding is used to differentiate slashes used in the URL path from slashes used to delimit folder names. Do you know if your blob container is HNS enabled?

**Comment by egdedudette** [comment]
Well this is so annoying. Before I could just copy the URL and now I have to remove the undesirable characters and replace them! SERIOUSLY! Great improvement!!!!

**Comment by egdedudette** [comment]
How do you check HNS enabled?

**Comment by JasonYeMSFT** [comment]
Next to your storage account node we show (ADLS Gen2) if the storage account is HNS enabled. To summarize your concern, you would like to have the copy url give a url that contains the `?snapshot` parameter and without the `%2F` encoded characters, right?

**Comment by craxal** [comment]
@egdedudette The encoded slashes should have no impact on whether they work or. The blob service should know how to handle them.

Hierarchical namespace (HNS) is enabled when you create the storage account containing the blob container. You can tell if the account has HNS enabled if the storage account node as "(ADLS Gen2)" next to the name, like this:

![image](https://user-images.githubusercontent.com/3473026/81122751-77197200-8ee6-11ea-963d-0748284e70e6.png)

Before we get too far (I don't want to make any assumptions), are you trying to get the URL of a blob or a blob snapshot? When you try to use the URL

**Comment by egdedudette** [comment]
if says (Key)  and we have not changed anything and it was working before as I mentioned. So obviously something is different and it is the encoding of the /.

**Comment by craxal** [comment]
@egdedudette OK, I copied a URL that has the %2F in it, but it seems to work just fine on our end. Don't worry, we'll get to the bottom of this! To do that, we'll need your help in figuring out exactly what you're doing to see what we may be doing wrong on our end. Once we can figure out how the problem occurs, we can figure out how to fix it.

First, the "(Key)" part tells me that this is a storage account you connected to manually using an account name and key. In other words, you did not need to sign in using an Azure account. Does that sound correct?

Second, I assume you're using the Copy URL menu command, yes? Or are you clicking on Properties?

Third, can you share with me the exact steps you take from starting up Storage Explorer all the way up to the point you are getting your URL?

---

Here are the steps I am taking, so let me know if I did something different than what you did.

1. Launch Storage Explorer and click the plug icon to open the Connect dialog.
2. Select "Use a storage account name and key", click Next.
3. Enter the account name (let's call it "myaccount") and key, but I don't change anything else, then click Next.
4. I look at the Summary to make sure everything looks correct, then I click Connect.
5. I now see a new node in the tree to the left: **Local & Attached** > **Storage Accounts** > **myaccount**. I expand that node to reveal **Blob Containers**, and I expand that node to see all of my blob containers. I click on one of those container (let's call it "mycontainer") to open the explorer on the right.
7. I look at the Properties tab below the tree on the left and notice that "mycontainer" has a Public Read Access property that says "blob". This tells me that anyone can read blobs in my container. _Your blob container might have a different value_.
6. I open the folder containing my blob, right-click the blob (in my case, a image named "picture.png"), and select Copy URL. I paste the URL into a word processor and notice that the URL has the encoded slashes:
   https://myaccount.blob.core.windows.net/mycontainer/folder%2Fpicture.png
7. To test my URL, I paste it into my browser and press Enter.

At this point, I am able to see my picture in my browser. I would expect the URL to work, because my container has "blob" public read access. If my container's public read access is "off", then I get this error page, as I would expect:

```
This XML file does not appear to have any style information associated with it. The document tree is shown below.
<Error>
<Code>ResourceNotFound</Code>
<Message>
The specified resource does not exist. RequestId:b9339998-e01e-0073-4633-2340b5000000 Time:2020-05-05T23:20:04.8250191Z
</Message>
</Error>
```

Do my step follow yours accurately, or did I do something majorly different than you?

**Comment by egdedudette** [comment]
First, the "(Key)" part tells me that this is a storage account you connected to manually using an account name and key. In other words, you did not need to sign in using an Azure account. Does that sound correct?
Correct

OK. I just noticed in the property window that it says:
HNS Enabled false  (which was like that before in the older version) settings are the same now.

I did exactly as you told me and I get the same result with the pasted / as %2F and my browser cannot read it properly.
I am testing Animate CC media. It shows the stage but with a broken icon. It says 
ReferenceError: AdobeAn is not defined. 
So basically it does not read the html correctly since it is where the file is defined.

**Comment by craxal** [comment]
@egdedudette I don't think that's a URL problem then, if your browser loads still loads the file, albeit with some bugs. I don't know a lot about Animate, but if you're creating HTML files, can you download the file and check the HTML to see if it looks correct? You should be able to open the original and downloaded files in your browser to check if they both perform the same way. Usually you just need a URL that looks like this:

file:///C:/path/to/your/file.html (if you're on Windows, and your file is on the C drive).

**Comment by egdedudette** [comment]
But the same file runs properly with the previous version of Azure.
It also works fine from my web site.
Animate cannot run from local folders because of security issues. Actually it loads properly but click events are disabled by the browser.
So not it is not a problem with the animate published HTML5 otherwise it would never run.
I copied multiples URL from files we had tested before with the previous version of Azure and with the new URL from the  new Azure they do not work. When I go back to the old version and copy the URL from there, it works.

**Comment by egdedudette** [comment]
Files copied are the index.html of published Animate CC media
Files are in blob container

recap:

AZURE PREVIOUS VERSION
- copy URL button on ribbon - nice touch
- / between folders
- HND enable = false  
- lease status: unlocked  
- files run without any problem

AZURE CURRENT VERSION
- copy URL button is on file right click
- %2F between folders
- HND enable = false  
- lease status: unlocked  
- files do not run unless %2F instances are replaced by /

Note: the file loads but fails as it seems to not read the html properly to get the media ID number which does not even makes senses since it reads it when the / is used.

**Comment by craxal** [comment]
@egdedudette Just to confirm I'm understanding you correctly:
- Your blob container has a public read access of "blob".
- In previous versions of Storage Explorer you get a URL with no %2F encoded slashes. This URL works as expected.
- In 1.13.0, you get a URL with %2F encoded slashes. This URL _does not_ work.
- In 1.13.0, if you manually replace %2F with "/", the URL _does_ work.

**Comment by egdedudette** [comment]
yes, exactly. If after placing the / it works as expected or as in previous version.

**Comment by egdedudette** [comment]
and thank you for looking into it.

**Comment by v-limu** [comment]
Verified on build [20200514.2](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=3726966&view=results). This issue has been fixed. So close it.

**Comment by v-xianya** [comment]
This issue reproduce on build master/20200529.3. So we re-open it.

**Comment by v-limu** [comment]
Verified on build 20200602.3. This issue has been fixed. So close it.
